overlord/devicestate/devicestate.go

```diff
@@ -478,25 +482,76 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 		if firstInstallInChain == nil {
 			firstInstallInChain = installFirst
 		}
-	}
-	// Make sure the first install waits for the last download. With this
-	// our (simplified) wait chain looks like:
-	// download1 <- verify1 <- download2 <- verify2 <- download3 <- verify3 <- install1 <- install2 <- install3
+		// download is always a first task of the 'download' phase
+		snapSetupTasks = append(snapSetupTasks, downloadLast.ID())
+	}
+	// Make sure the first install waits for the recovery system (only in
+	// UC20) which waits for the last download. With this our (simplified)
+	// wait chain looks like this:
+	//
+	// download1
+	//   ^- verify1
+	//        ^- download2
+	//             ^- verify2
+	//                  ^- download3
+	//                       ^- verify3
+	//                            ^- recovery (UC20)
+	//                                 ^- install1
+	//                                      ^- install2
+	//                                           ^- install3
 	if firstInstallInChain != nil && lastDownloadInChain != nil {
 		firstInstallInChain.WaitFor(lastDownloadInChain)
 	}
 
+	recoverySetupTaskID := ""
+	if current.Grade() != asserts.ModelGradeUnset && new.Grade() != asserts.ModelGradeUnset {
```

Comment:
conceptually we could check just for new here, no? the caller check for the policy of what is possible